pyccl/tests/test_power.py

```diff
@@ -187,89 +211,364 @@ def test_power_sigma8norm_norms_consistent(tf):
+            Omega_c=0.27, Omega_b=0.05, h=0.7,
+            n_s=0.965, sigma8=0.8,
+            nonlinear_model='halofit')
+
+    with pytest.raises(KeyError):
+        ccl.CosmologyCalculator(
+            Omega_c=0.27, Omega_b=0.05, h=0.7,
+            n_s=0.965, sigma8=0.8,
+            pk_linear={'a': a_arr, 'k': k_arr,
+                       'delta_matter:delta_matter': pkl_arr},
+            nonlinear_model={'y:z': 'halofit'})
+
+    with pytest.raises(ValueError):
+        ccl.CosmologyCalculator(
+            Omega_c=0.27, Omega_b=0.05, h=0.7,
+            n_s=0.965, sigma8=0.8,
+            pk_linear={'a': a_arr, 'k': k_arr,
+                       'delta_matter:delta_matter': pkl_arr},
+            nonlinear_model={'delta_matter:delta_matter': None})
+
+    with pytest.raises(KeyError):
+        ccl.CosmologyCalculator(
+            Omega_c=0.27, Omega_b=0.05, h=0.7,
+            n_s=0.965, sigma8=0.8,
+            pk_linear={'a': a_arr, 'k': k_arr,
+                       'delta_matter:delta_matter': pkl_arr},
+            nonlinear_model={'delta_matter:delta_matter': 'hmcode'})
+
 
 def test_input_nonlin_raises():
-    cosmo_input = ccl.Cosmology(Omega_c=0.27, Omega_b=0.05, h=0.7, n_s=0.965,
-                                A_s=2e-9)
+    cosmo = ccl.Cosmology(Omega_c=0.27, Omega_b=0.05, h=0.7,
+                          n_s=0.965, sigma8=0.8,
+                          transfer_function='bbks')
+    a_arr = np.linspace(0.1, 1.0, 50)
+    k_arr = np.logspace(np.log10(2e-4), np.log10(1), 1000)
+    pkl_arr = np.array([ccl.power.linear_matter_power(cosmo, k_arr, a)
+                        for a in a_arr])
+    pk_arr = np.array([ccl.power.nonlin_matter_power(cosmo, k_arr, a)
+                       for a in a_arr])
+
+    # Not a dictionary
+    with pytest.raises(TypeError):
+        ccl.CosmologyCalculator(
+            Omega_c=0.27, Omega_b=0.05, h=0.7,
+            n_s=0.965, sigma8=0.8,
+            pk_nonlin=np.pi)
+
+    # k not present
+    with pytest.raises(ValueError):
+        ccl.CosmologyCalculator(
+            Omega_c=0.27, Omega_b=0.05, h=0.7,
+            n_s=0.965, sigma8=0.8,
+            pk_nonlin={'a': a_arr, 'kk': k_arr,
+                       'delta_matter_y_delta_matter': pk_arr})
+
+    # delta_matter:delta_matter not present
+    with pytest.raises(ValueError):
+        ccl.CosmologyCalculator(
+            Omega_c=0.27, Omega_b=0.05, h=0.7,
+            n_s=0.965, sigma8=0.8,
+            pk_nonlin={'a': a_arr, 'k': k_arr,
+                       'delta_matter_y_delta_matter': pk_arr})
+
+    # Non-parsable power spectrum
     with pytest.raises(ValueError):
-        cosmo_input._set_nonlin_power_from_arrays()
+        ccl.CosmologyCalculator(
+            Omega_c=0.27, Omega_b=0.05, h=0.7,
+            n_s=0.965, sigma8=0.8,
+            pk_nonlin={'a': a_arr, 'k': k_arr,
+                       'delta_matter:delta_matter': pk_arr,
+                       'a_y_b': pk_arr})
```

Comment:
should this be `:`?

Reply:
We want to test that it fails. Changed it to `;`

Reply:
ahhhh great!